### PR TITLE
Remove runtime compilation from MVC.Razor

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionInvokerProviderTest.cs [hunk 1]
```diff
@@ -28,15 +27,13 @@ namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
...
-        private readonly IHostingEnvironment _hostingEnvironment = Mock.Of<IHostingEnvironment>(e => e.ContentRootPath == "BasePath");
-
         [Fact]
         public void OnProvidersExecuting_WithEmptyModel_PopulatesCacheEntry()
         {
             // Arrange
             var descriptor = new PageActionDescriptor
             {
-                RelativePath = "Path1",
+                RelativePath = "/Path1",
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionInvokerProviderTest.cs]: As mentioned earlier, path w/o leading slash should still work. I don't mean this change should be undone (was previously testing something that should be impossible -- invalid `PageActionDescriptor`s) but please confirm lookups are tested (and succeed) using a path w/o the initial slash.

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorFileHierarchy.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        public static IEnumerable<string> FindViewImports(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(path));
+            }
+
+            if (path[0] != '/')
+            {
+                throw new ArgumentException(Resources.RazorProject_PathMustStartWithForwardSlash, nameof(path));
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Razor/RazorFileHierarchy.cs]: The current code does this: 

* https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Language/RazorProject.cs#L62
* https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Language/RazorProject.cs#L112-L125

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs [hunk 3]
```diff
@@ -88,9 +88,6 @@ internal static void AddServices(IServiceCollection services)
             // Action description and invocation
             services.TryAddEnumerable(
                 ServiceDescriptor.Singleton<IActionDescriptorProvider, PageActionDescriptorProvider>());
-            services.TryAddSingleton<IActionDescriptorChangeProvider, PageActionDescriptorChangeProvider>();
-            services.TryAddEnumerable(
-                ServiceDescriptor.Singleton<IPageRouteModelProvider, RazorProjectPageRouteModelProvider>());
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcCoreBuilderExtensions.cs]: Build time compiled pages are discovered by `CompiledPageRouteModelProvider`

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionInvokerProviderTest.cs (hunk 1 above)
Comment (pranavkm) [reply test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageActionInvokerProviderTest.cs]: `VirtualRazorProjectFileSystem` that's used in tests supported relative paths without leading slashes, but the default implementation did not. This wasn't a valid test case

File: test/WebSites/RazorWebSite/Controllers/PartialViewEngineController.cs [hunk 4]
```diff
@@ -14,7 +14,7 @@ public IActionResult ViewWithoutLayout()
 
         public IActionResult ViewWithFullPath()
         {
-            return PartialView("/Views/ViewEngine/ViewWithFullPath.rzr");
+            return PartialView("/Views/ViewEngine/ViewWithFullPath.cshtml");
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply test/WebSites/RazorWebSite/Controllers/PartialViewEngineController.cs]: It does, but it's not an meaningful scenario. Plus changing the project to support arbitrary extension would be to test an Sdk feature, not what this scenario purports to do

**Comment by pranavkm** [comment]
🆙 :date

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
cc @natemcmaster \ @ryanbrandenburg

**Pull request merged**